### ISSUE TITLE
linux-builtin installition no longer work in 2.1.13

### ISSUE THREAD

**Issue by RinCat** [openzfs/zfs#15338]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Gentoo
Distribution Version	| N/A
Kernel Version	| 6.5.5
Architecture	| amd64
OpenZFS Version	| 2.1.13

### Describe the problem you're observing

`linux-builtin` install no longer work.
There are no errors during any steps, but the `/dev/zfs` is missing and `zfs` / `zpool` reported zfs module not loaded.
`dmesg` shows no zfs related logs.

### Describe how to reproduce the problem

1. Configure zfs with `--enable-linux-builtin`
2. Install it use `./copy-builtin /usr/src/linux`
3. Build kernel `.config` with `CONFIG_ZFS=y`
4. Reboot

### Include any warning/errors/backtraces from the system logs

No log, kernel does not see the builtin module exists.

ZFS files exist:
```
find /usr/src/linux/ -iname "*zfs*"

/usr/src/linux/fs/zfs
/usr/src/linux/fs/zfs/zcommon/zfs_comutil.c
/usr/src/linux/fs/zfs/zcommon/zfs_namecheck.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher.c
/usr/src/linux/fs/zfs/zcommon/zfs_prop.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher_superscalar4.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher_superscalar.c
/usr/src/linux/fs/zfs/zcommon/zfs_deleg.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher_avx512.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher_sse.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher_intel.c
/usr/src/linux/fs/zfs/zcommon/zfs_fletcher_aarch64_neon.c
/usr/src/linux/fs/zfs/zstd/zfs_zstd.c
/usr/src/linux/fs/zfs/lua/README.zfs
/usr/src/linux/fs/zfs/zfs
/usr/src/linux/fs/zfs/zfs/zfs_sa.c
/usr/src/linux/fs/zfs/zfs/zfs_quota.c
/usr/src/linux/fs/zfs/zfs/zfs_fuid.c
/usr/src/linux/fs/zfs/zfs/edonr_zfs.c
/usr/src/linux/fs/zfs/zfs/zfs_byteswap.c
/usr/src/linux/fs/zfs/zfs/zfs_rlock.c
/usr/src/linux/fs/zfs/zfs/skein_zfs.c
/usr/src/linux/fs/zfs/zfs/zfs_log.c
/usr/src/linux/fs/zfs/zfs/zfs_ioctl.c
/usr/src/linux/fs/zfs/zfs/zfs_fm.c
/usr/src/linux/fs/zfs/zfs/zfs_onexit.c
/usr/src/linux/fs/zfs/zfs/zfs_vnops.c
/usr/src/linux/fs/zfs/zfs/zfs_replay.c
/usr/src/linux/fs/zfs/zfs/zfs_ratelimit.c
/usr/src/linux/fs/zfs/os/linux/zfs
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_dir.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_sysfs.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_vfsops.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_ctldir.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_racct.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_znode.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_vnops_os.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_acl.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_uio.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_ioctl_os.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_debug.c
/usr/src/linux/fs/zfs/os/linux/zfs/zfs_file_os.c
/usr/src/linux/fs/zfs/os/freebsd/zfs
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_debug.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_ioctl_os.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_file_os.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_dir.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_vnops_os.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_znode.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_racct.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_vfsops.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_ctldir.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_acl.c
/usr/src/linux/fs/zfs/os/freebsd/zfs/zfs_ioctl_compat.c
/usr/src/linux/include/zfs
/usr/src/linux/include/zfs/libzfsbootenv.h
/usr/src/linux/include/zfs/libzfs_core.h
/usr/src/linux/include/zfs/libzfs.h
/usr/src/linux/include/zfs/os/freebsd/spl/sys/trace_zfs.h
/usr/src/linux/include/zfs/os/freebsd/zfs
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_ctldir.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_dir.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_context_os.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_bootenv_os.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_vfsops_os.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_znode_impl.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_ioctl_compat.h
/usr/src/linux/include/zfs/os/freebsd/zfs/sys/zfs_vnops_os.h
/usr/src/linux/include/zfs/os/linux/zfs
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_dir.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_ctldir.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/trace_zfs.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_znode_impl.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_vfsops_os.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_vnops_os.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_bootenv_os.h
/usr/src/linux/include/zfs/os/linux/zfs/sys/zfs_context_os.h
/usr/src/linux/include/zfs/zfs_deleg.h
/usr/src/linux/include/zfs/libzfs_impl.h
/usr/src/linux/include/zfs/zfs_namecheck.h
/usr/src/linux/include/zfs/zfs_gitrev.h
/usr/src/linux/include/zfs/zfs_prop.h
/usr/src/linux/include/zfs/zfs_config.h
/usr/src/linux/include/zfs/sys/zfs_acl.h
/usr/src/linux/include/zfs/sys/zfs_rlock.h
/usr/src/linux/include/zfs/sys/zfs_quota.h
/usr/src/linux/include/zfs/sys/zfs_sa.h
/usr/src/linux/include/zfs/sys/zfs_delay.h
/usr/src/linux/include/zfs/sys/zfs_debug.h
/usr/src/linux/include/zfs/sys/zfs_vnops.h
/usr/src/linux/include/zfs/sys/zfs_ioctl_impl.h
/usr/src/linux/include/zfs/sys/zfs_vfsops.h
/usr/src/linux/include/zfs/sys/zfs_fuid.h
/usr/src/linux/include/zfs/sys/zfs_ratelimit.h
/usr/src/linux/include/zfs/sys/zfs_refcount.h
/usr/src/linux/include/zfs/sys/zfs_project.h
/usr/src/linux/include/zfs/sys/zfs_context.h
/usr/src/linux/include/zfs/sys/zfs_ioctl.h
/usr/src/linux/include/zfs/sys/zfs_onexit.h
/usr/src/linux/include/zfs/sys/zfs_sysfs.h
/usr/src/linux/include/zfs/sys/fs/zfs.h
/usr/src/linux/include/zfs/sys/fm/fs/zfs.h
/usr/src/linux/include/zfs/sys/zfs_znode.h
/usr/src/linux/include/zfs/sys/zfs_file.h
/usr/src/linux/include/zfs/sys/zfs_stat.h
/usr/src/linux/include/zfs/sys/zfs_bootenv.h
/usr/src/linux/include/zfs/sys/zfs_racct.h
/usr/src/linux/include/zfs/zfs_comutil.h
/usr/src/linux/include/zfs/zfs_fletcher.h
/usr/src/linux/include/config/ZFS
```

```
grep ZFS /usr/src/linux/.config

CONFIG_ZFS=y
```

**Comment by siliciferous** [comment]
I can confirm this issue; also on Gentoo, also with kernel version 6.5.5 (sys-kernel/gentoo-sources) using OpenZFS 2.1.13 (sys-fs/zfs-kmod). OpenZFS 2.1.13's _./copy-builtin_ works as it always has with kernel 6.1.53.

_copy-builtin_ displays no errors when pointed at 6.5.5's source tree.
ZFS appears in menuconfig as it should after _copy-builtin_ is executed against it.
When building 6.5.5 there is a conspicuous absence of zfs entries in the kernel build log.

**Comment by Lisio** [comment]
Same issue on Ubuntu.

**Comment by daniejstriata** [comment]
Version 2.1.13 is causing havoc on my RHEL hosts. Most of my upgrades fail to build the kernel module.

**Comment by spillner** [comment]
The problem is that `copy-builtin` tries to add the `obj-$(CONFIG-ZFS) += zfs/` line after finding an `endif` in `fs/Makefile`, but Linux 6.5.5 changed `fs/Makefile` to replace that old `endif` with separate `obj-$(CONFIG_BLOCK)`, `obj-$CONFIG_LEGACY_DIRECT_IO)`, etc. lines.

There's no reason that the ZFS target needs to be inserted after a specific line in `fs/Makefile`; it might as well just be echoed onto the very end of the file.  As a temporary fix, you can manually insert the line into `fs/Makefile`, or change the pattern passed to the `add_after` line in `copy-builtin` to something more enduring, like `#`.

**Comment by pega2k** [comment]
With the ZFS release version 2.2.0 this is working again.

**Comment by IvanVolosyuk** [comment]
Can be closed now I guess, fixed in 2.1.14
